src/codec.rs

```diff
@@ -637,7 +641,74 @@ pub(crate) fn encode_slice_no_len<T: Encode, W: Output + ?Sized>(slice: &[T], de
 	}
 }
 
-/// Decode the slice (without prepended the len).
+/// Decode the array.
+///
+/// This is equivalent to decoding all the element one by one, but it is optimized for some types.
+#[inline]
+pub(crate) fn decode_array<I: Input, T: Decode, const N: usize>(input: &mut I) -> Result<[T; N], Error> {
+	#[inline]
+	fn general_array_decode<I: Input, T: Decode, const N: usize>(input: &mut I) -> Result<[T; N], Error> {
+		let mut uninit = <MaybeUninit<[T; N]>>::uninit();
+		// The following line coerces the pointer to the array to a pointer
+		// to the first array element which is equivalent.
+		let mut ptr = uninit.as_mut_ptr() as *mut T;
+		for _ in 0..N {
+			let decoded = T::decode(input)?;
+			// SAFETY: We do not read uninitialized array contents
+			//         while initializing them.
+			unsafe {
+				ptr::write(ptr, decoded);
+			}
+			// SAFETY: Point to the next element after every iteration.
+			//         We do this N times therefore this is safe.
+			ptr = unsafe { ptr.add(1) };
+		}
+		// SAFETY: All array elements have been initialized above.
+		let init = unsafe { uninit.assume_init() };
+		Ok(init)
+	}
+
+	macro_rules! decode {
+		( u8 ) => {{
+			let mut array: ManuallyDrop<[u8; N]> = ManuallyDrop::new([0; N]);
+			input.read(&mut array[..])?;
+			let ref_typed: &[T; N] = unsafe { mem::transmute(&array) };
+			let typed: [T; N] = unsafe { ptr::read(ref_typed) };
+			Ok(typed)
+		}};
+		( i8 ) => {{
+			let mut array: ManuallyDrop<[i8; N]> = ManuallyDrop::new([0; N]);
+			let bytes = unsafe { mem::transmute::<&mut [i8], &mut [u8]>(&mut array[..]) };
+			input.read(bytes)?;
+
+			let ref_typed: &[T; N] = unsafe { mem::transmute(&array) };
+			let typed: [T; N] = unsafe { ptr::read(ref_typed) };
+			Ok(typed)
```

Comment:
```suggestion
			Ok(unsafe { mem::transmute(array.into_inner()) })
```

Why can we not do this?

Reply:
Because we can't convert due to this issue https://github.com/rust-lang/rust/issues/61956

Reply:
Ahh okay. Why don't we need to drop `array`? I assume this will then be done when dropping `typed`?

Reply:
`array` and `typed` use the same memory. We return `typed` from the functions, so the compiler will not call `drop` on that variable. But it will drop `array`. So to avoid dropping of shared memory we don't need to drop `array`=)

Reply:
Maybe you could add some comments explaining this to people looking at this code in the future.

Reply:
> Maybe you could add some comments explaining this to people looking at this code in the future.

Sounds like a good idea :)